controller/volume_controller.go

```diff
@@ -1553,6 +1639,269 @@ func getRebuildingReplicaCount(e *longhorn.Engine) int {
+		if i == item {
+			return true
+		}
+	}
+	return false
+}
+
+func (vc *VolumeController) replicaAutoBalanceZoneCount(v *longhorn.Volume, rs map[string]*longhorn.Replica) (int, map[string][]string, error) {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceType", "zone")
+
+	readyNodes, err := vc.ds.ListReadyAndSchedulableNodes()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var usedZones []string
+	var usedNodes []string
+	zoneDuplicates := make(map[string][]string)
+	for _, r := range rs {
+		if r.Status.CurrentState != types.InstanceStateRunning {
+			continue
+		}
+
+		node, exist := readyNodes[r.Spec.NodeID]
+		if !exist {
+			log.Debugf("Replica %v on node %v not count for auto-balance, could get evicted", r.Spec.NodeID, r.Name)
+			continue
+		}
+
+		nZone := node.Status.Zone
+		_, exist = zoneDuplicates[nZone]
+		if exist {
+			zoneDuplicates[nZone] = append(zoneDuplicates[nZone], r.Name)
+		} else {
+			zoneDuplicates[nZone] = nil
+			usedZones = append(usedZones, nZone)
+		}
+		if !contains(usedNodes, r.Spec.NodeID) {
+			usedNodes = append(usedNodes, r.Spec.NodeID)
+		}
+	}
+	log.Debugf("Found %v use zones %v", len(usedZones), usedZones)
+	log.Debugf("Found %v use nodes %v", len(usedNodes), usedNodes)
+	if v.Spec.NumberOfReplicas == len(zoneDuplicates) {
+		log.Debugf("Balanced, %v volume replicas are running on different zones", v.Spec.NumberOfReplicas)
+		return 0, zoneDuplicates, nil
+	}
+
+	ei, err := vc.getEngineImage(v.Status.CurrentImage)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	unusedZone := make(map[string][]string)
+	for nodeName, node := range readyNodes {
+		if contains(usedZones, node.Status.Zone) {
+			log.Debugf("Cannot use node %v in zone '%v', have running replica", nodeName, node.Status.Zone)
+			continue
+		}
+
+		if contains(usedNodes, nodeName) {
+			log.Debugf("Cannot use node %v, have running replica", nodeName)
+			continue
+		}
+
+		if !node.Spec.AllowScheduling {
+			log.Debugf("Cannot use node %v, does not allow scheduling", nodeName)
+			continue
+		}
+
+		if isReady, _ := vc.ds.CheckEngineImageReadiness(ei.Spec.Image, nodeName); !isReady {
+			log.Debugf("Cannot use node %v, Engine image is not ready", nodeName)
+			continue
+		}
+
+		unusedZone[node.Status.Zone] = append(unusedZone[node.Status.Zone], nodeName)
+	}
+	if len(unusedZone) == 0 {
+		log.Debug("Balanced, all ready zones are used by this volume")
+		return 0, zoneDuplicates, err
+	}
+
+	unevenCount := v.Spec.NumberOfReplicas - len(zoneDuplicates)
+	unusedCount := len(unusedZone)
+	adjustCount := 0
+	if unusedCount < unevenCount {
+		adjustCount = unusedCount
+	} else {
+		adjustCount = unevenCount
+	}
+	log.Debugf("Found %v zone available for auto-balance duplicates in %v", adjustCount, zoneDuplicates)
+
+	return adjustCount, zoneDuplicates, err
+}
+
+func (vc *VolumeController) replicaAutoBalanceNodeCount(v *longhorn.Volume, rs map[string]*longhorn.Replica) (int, map[string][]string, error) {
```

Comment:
Change to `getReplicaCountForAutoBalanceNode`

Reply:
Addressed.